### PR TITLE
Adds SensorTypes and GoalSensors

### PR DESCRIPTION
**Pull request by awjuliani**
### Proposed change(s)

Replicates changes from https://github.com/Unity-Technologies/ml-agents/tree/develop-taggedobservations with `TaskSensor` changes to `GoalSensor`. This was easier than cherry picking from the original branch, as it was quite outdated after the hybrid action changes.

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Sensors/ISensor.cs [hunk 1]
```diff
@@ -16,6 +16,28 @@ public enum SensorCompressionType
         PNG
     }
 
+    /// <summary>
```
Comment (andrewcoh) [inline com.unity.ml-agents/Runtime/Sensors/ISensor.cs:19]: Whether or not this enum makes it into master will ultimately be up to the toolkit pod but for our purposes whatever works.

File: ml-agents-envs/mlagents_envs/base_env.py [hunk 2]
```diff
@@ -367,10 +374,13 @@ class BehaviorSpec(NamedTuple):
     - observation_shapes is a List of Tuples of int : Each Tuple corresponds
     to an observation's dimensions. The shape tuples have the same ordering as
     the ordering of the DecisionSteps and TerminalSteps.
+    - sensor_types is a List of SensorTypes, each corresponding to the type of
+    sensor (i.e. observation, goal, etc).
     - action_spec is an ActionSpec NamedTuple
     """
 
     observation_shapes: List[Tuple]
+    sensor_types: List[SensorType]
```
Comment (andrewcoh) [inline ml-agents-envs/mlagents_envs/base_env.py:383]: Looks like we need to add on ObsSpec tuple to contain sensor types and obs shapes

**Review by andrewcoh — APPROVED** [review]

File: com.unity.ml-agents/Runtime/Sensors/ISensor.cs (hunk 1 above)
Comment (awjuliani) [reply com.unity.ml-agents/Runtime/Sensors/ISensor.cs:19]: Agreed. This is just for testing purposes.

File: ml-agents-envs/mlagents_envs/base_env.py (hunk 2 above)
Comment (awjuliani) [reply ml-agents-envs/mlagents_envs/base_env.py:383]: Oh yeah. I was using the old code, but it makes sense to do this when we want to make it official.

**Pull request merged**